### PR TITLE
#126 Implementing SSH private key connection to staging

### PR DESCRIPTION
**Pull request by bkuzmic**
For #126:

- Implementing reading the credentials.json and using private key and username to connect to staging server, transfer files and execute commands
- leaving puzzle to finish adding cron scripts and creating curl requests to test that application is running ok

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#223](https://github.com/timelesslounge/timelessis/issues/223) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#223](https://github.com/timelesslounge/timelessis/issues/223) is assigned to @vryazanov[/z](https://www.0crat.com/u/vryazanov), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/061f1af7-dd61-4013-b459-dc14059d6dfc); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: deploy.sh [hunk 1]
```diff
@@ -1,24 +1,27 @@
...
 echo "-- Copy application code to staging server"
-scp -r . user@staging-server:/app
+scp -i $STAGING_KEY -r . $STAGING_USER@$STAGING_SERVER:/app
 
 # add scripts in cron (like the one created in #47)
 # verify the webapplication is running
-ssh user@staging-server << EOF
-  echo "-- Creating database user: timless_user"
+ssh -i $STAGING_KEY $STAGING_USER@$STAGING_SERVER << EOF
+  echo "-- Creating database user: timeless_user"
```
Comment (vryazanov) [inline deploy.sh]: should we take db user from credentials.json as well?

Comment (bkuzmic) [reply deploy.sh]: @vryazanov Yes, of course, make sense.

**Comment by bkuzmic** [comment]
@vryazanov Please take a look again.

**Comment by vryazanov** [comment]
@rultor merge

File: deploy.sh [hunk 2]
```diff
@@ -1,38 +1,45 @@
...
     INHERIT
     LOGIN
-    ENCRYPTED PASSWORD 'timeless_pwd';"
+    ENCRYPTED PASSWORD '$PG_STAGING_PASS';"
   echo "-- Creating database: timelessdb_dev"  
   sudo -u postgres psql -c "CREATE DATABASE timelessdb_dev;"
   echo "-- Creating database: timelessdb_test"  
   sudo -u postgres psql -c "CREATE DATABASE timelessdb_test;"
   echo "-- REPLACE: add scripts to cron"
   cd /app
```
Comment (emilianodellacasa) [inline deploy.sh]: @bkuzmic We need also to update external libraries, right?

File: deploy.sh [hunk 3]
```diff
@@ -1,38 +1,45 @@
 #!/bin/sh
 
 # Scripts to deploy application to staging server
-# @todo #45:30min Continue implementing deploy scripts. Find out user 
-#  and domain of the staging server and replace scp and ssh with correct data.
-#  Add scripts in cron (like the one created in #47). Verify the web 
-#  application is running (execute couple fo curl requests).
+# @todo #126:30min Finish implementing deploy scripts. Add scripts in cron 
+#  (like the one created in #47). Verify the web application is running - 
+#  execute couple of curl requests.
```
Comment (emilianodellacasa) [inline deploy.sh]: @bkuzmic I would also add a puzzle to add another parameter to decide in which environment to deploy, either 'staging' or 'production'

**Comment by emilianodellacasa** [comment]
@bkuzmic A couple of comments for you

File: deploy.sh [hunk 4]
```diff
@@ -1,38 +1,45 @@
...
 
 echo "-- Run tests"
```
Comment (emilianodellacasa) [inline deploy.sh]: @bkuzmic As I understand, these script does not install the complete environment (python, postgres and so on ) and that is fine, but I would add a puzzle to describe the installation process for the staging/production servers.

File: deploy.sh (hunk 2 above)
Comment (bkuzmic) [reply deploy.sh]: @emilianodellacasa What do you mean by external libraries? Do you mean scripts, like backup or something else?

File: deploy.sh (hunk 3 above)
Comment (bkuzmic) [reply deploy.sh]: @emilianodellacasa Sure, I'll do that.

File: deploy.sh (hunk 2 above)
Comment (emilianodellacasa) [reply deploy.sh]: @bkuzmic No, I mean flask libraries. I am just asking to add a `pip install -r requirements.txt` here

File: deploy.sh (hunk 4 above)
Comment (emilianodellacasa) [reply deploy.sh]: @bkuzmic In the same puzzle, explain how to call rultor to deploy 

Ex.

`@rultor deploy 0.0.1 staging`

**Comment by bkuzmic** [comment]
@emilianodellacasa Please take a look now.

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 23 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @vryazanov[/z](https://www.0crat.com/u/vryazanov), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#223](https://github.com/timelesslounge/timelessis/issues/223) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@vryazanov According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only one issue was found during code review. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by vryazanov** [comment]
@ypshenychka I confirm

**Comment by ypshenychka** [comment]
@vryazanov thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +20 point(s) just awarded to @vryazanov[/z](https://www.0crat.com/u/vryazanov)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**